### ISSUE TITLE
[BUG] final_layernorm gradient mismatch when TP+SP enabled

### ISSUE THREAD

**Issue by i-love-megatron** [NVIDIA/Megatron-LM#1446]
**Describe the bug**
In `pretrain_gpt.py`, when both tensor parallel and sequence parallel are enabled, the `main_grad` of parameters in `final_layernorm` are different on different TP ranks before `optimizer.step()`.

**To Reproduce**
Run `pretrain_gpt.py` with both `--tensor-model-parallel-size` larger than 1 and `--sequence-parallel` opened.

**Expected behavior**
Layernorms parameters are partitioned, so their gradients are expected to be the same before `optimizer.step()` on different parallel ranks.

**Stack trace/logs**
I put the values of final_layernorm.weight.main_grad below:
TP rank 0 has value:
tensor([2.9907e-03, -3.1120e-04,  9.9897e-05,  ...,  1.7762e-04, -7.4244e-04, -4.4298e-04]
TP rank 1:
tensor([2.9755e-03, -6.2752e-04,  2.9802e-05,  ...,  2.2030e-04, -9.9659e-04, -5.0163e-04]
TP rank 2:
tensor([3.0289e-03, -5.4169e-04, -1.4603e-05,  ...,  2.4462e-04, -8.9359e-04, -3.8719e-04]
TP rank 3:
tensor([1.9131e-03, -4.3297e-04,  6.0797e-06,  ...,  3.1662e-04, -8.9264e-04, -5.1117e-04]

**Environment (please complete the following information):**
 - Megatron-LM commit ID: The latest commit from main branch, fdfcef87004205eca69ea462b375578185ea01f6
 - PyTorch version: 2.4.1+cu124
 - CUDA version: 12.4
 - NCCL version: 2.20.5

**Proposed fix**
N/A

**Additional context**
N/A

**Comment by i-love-megatron** [comment]
Solved by https://github.com/NVIDIA/TransformerEngine/pull/1528.
